<h1>stenglein-lab/tick_surveillance pipeline parameters</h1>
<p>A pipeline to analyze amplicon sequencing data for tick-borne pathogens</p>
<h2>Primer and adapter trimming options</h2>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Type</th>
<th>Default</th>
<th>Required</th>
<th>Hidden</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>primers</code></td>
<td>A file containing the names and sequences of primers used to amplify targets.</td>
<td><code>string</code></td>
<td>refseq/primers.tsv</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>post_trim_min_length</code></td>
<td>After trimming of adapter and primer sequences, amplicons shorter than this length will be discarded.</td>
<td><code>integer</code></td>
<td>100</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>amplicon_primers_max_error_fraction</code></td>
<td>This specifies the error tolerance (fraction) used when searching for adapters sequences to trim. This value is passed to the 
the cutadapt -e parameter.</td>
<td><code>number</code></td>
<td>0.2</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>adapters_min_overlap</code></td>
<td>Used in trimming of Illumina adapter sequences.  Specifies the minimum length of overlap between adapter sequence and read 
sequence for trimming to occur.  Passed to cutadapt -O parameter.</td>
<td><code>integer</code></td>
<td>10</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<h2>Calling of positives and Surveillance Report</h2>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Type</th>
<th>Default</th>
<th>Required</th>
<th>Hidden</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>surveillance_columns</code></td>
<td>This file specifies what columns will be included in the surveillance report and optional default values.</td>
<td><code>string</code></td>
<td>refseq/surveillance_columns.txt</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>min_reads_for_positive_surveillance_call</code></td>
<td>The number of read pairs assigned to a particular target in order for that target to be called positive.</td>
<td><code>integer</code></td>
<td>50</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>max_blast_refseq_evalue</code></td>
<td>The maximum BLASTN e-value for an ASV to be initially assigned to a target sequence.  Final assignment will be based on 
additional criteria.</td>
<td><code>number</code></td>
<td>1e-10</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<h2>Input/output options</h2>
<p>Define where the pipeline should find input data and save output data.</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Type</th>
<th>Default</th>
<th>Required</th>
<th>Hidden</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>fastq_pattern</code></td>
<td>fastq files must match this pattern to be used as input to the pipeline.</td>
<td><code>string</code></td>
<td><em><em>R[12]</em></em>.fastq*</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>fastq_dir</code></td>
<td>A directory containing fastq files that will be input to the pipeline.</td>
<td><code>string</code></td>
<td>input/fastq/</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>outdir</code></td>
<td>The output directory where results will be saved. You have to use absolute paths to store on Cloud infrastructure.</td>
<td><code>string</code></td>
<td>./results/</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>targets</code></td>
<td>A file containing the target sequences and other information about these sequences.</td>
<td><code>string</code></td>
<td>refseq//targets.tsv</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>tracedir</code></td>
<td>Directory in which Nextflow logs and reports will be placed.</td>
<td><code>string</code></td>
<td>${params.outdir}/pipeline_info</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>log_outdir</code></td>
<td>Directory in which pipeline logs and reports will be placed.</td>
<td><code>string</code></td>
<td>${params.outdir}/log</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>output_prefix</code></td>
<td>Main pipeline output file names will contain this prefix.</td>
<td><code>string</code></td>
<td>Today's date in YYYY_MM_DD format</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<h2>Max job request options</h2>
<p>Set the top limit for requested resources for any single job.</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Type</th>
<th>Default</th>
<th>Required</th>
<th>Hidden</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>max_cpus</code></td>
<td>Maximum number of CPUs that can be requested for any single job. <details><summary>Help</summary><small>Use to set an 
upper-limit for the CPU requirement for each process. Should be an integer e.g. <code>--max_cpus 1</code></small></details></td>
<td><code>integer</code></td>
<td>16</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>max_memory</code></td>
<td>Maximum amount of memory that can be requested for any single job. <details><summary>Help</summary><small>Use to set an 
upper-limit for the memory requirement for each process. Should be a string in the format integer-unit e.g. <code>--max_memory 
'8.GB'</code></small></details></td>
<td><code>string</code></td>
<td>128.GB</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>max_time</code></td>
<td>Maximum amount of time that can be requested for any single job. <details><summary>Help</summary><small>Use to set an 
upper-limit for the time requirement for each process. Should be a string in the format integer-unit e.g. <code>--max_time 
'2.h'</code></small></details></td>
<td><code>string</code></td>
<td>240.h</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<h2>Generic options</h2>
<p>Less common options for the pipeline, typically set in a config file.</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Type</th>
<th>Default</th>
<th>Required</th>
<th>Hidden</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>help</code></td>
<td>Display help text.</td>
<td><code>boolean</code></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<h2>DADA2 options</h2>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Type</th>
<th>Default</th>
<th>Required</th>
<th>Hidden</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>dada_outdir</code></td>
<td>Directory in which DADA2 output files will be placed.</td>
<td><code>string</code></td>
<td>${params.outdir}/dada2</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<h2>BLASTing of unassigned sequences</h2>
<p>Parameters associated with BLASTing of unassigned sequences against the NCBI nt database for the purposes of taxonomic 
classification of non-target sequences.</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Type</th>
<th>Default</th>
<th>Required</th>
<th>Hidden</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>blast_unassigned_sequences</code></td>
<td>A flag to turn off or on this BLAST / assignment step.</td>
<td><code>boolean</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>max_blast_nt_evalue</code></td>
<td>The maximum BLAST e-value.</td>
<td><code>number</code></td>
<td>1e-10</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>blast_perc_identity</code></td>
<td>The minimum percent identity for BLAST hits to be considered.</td>
<td><code>number</code></td>
<td>70</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>blast_qcov_hsp_perc</code></td>
<td>The minimum query coverage percentage for BLAST hits to be considered.</td>
<td><code>number</code></td>
<td>70</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>local_nt_database_dir</code></td>
<td>The path to a directory containing a local copy of the NCBI nt database (or whatever database you wish to search).  Note that 
this is not the path of the database, but is the path of the directory containing the database.</td>
<td><code>string</code></td>
<td>None</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>local_nt_database_name</code></td>
<td>The name of the local BLAST database to be searched in this step.</td>
<td><code>string</code></td>
<td>nt</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>remote_blast_nt</code></td>
<td>Setting this parameter to true will run this BLAST search using NCBI's remote copy of the nt blast database.  This avoids 
needing to have a local copy of the nt database installed.  Note that this will be very slow.</td>
<td><code>boolean</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>blast_tax_dir</code></td>
<td>The directory containing local copies of the files contained in https://ftp.ncbi.nlm.nih.gov/blast/db/taxdb.tar.gz.  This is 
optional and if you don't specify it (recommended), the pipeine will download these automatically.</td>
<td><code>string</code></td>
<td>None</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<h2>Tree-building</h2>
<p>Parameters related to the construction of phylogenetic trees that include observed sequences and pre-defined reference 
sequences.</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Type</th>
<th>Default</th>
<th>Required</th>
<th>Hidden</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>make_trees</code></td>
<td>A flag to turn on or off this step.</td>
<td><code>boolean</code></td>
<td>True</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>tree_outdir</code></td>
<td>A directory in which tree files will be placed.</td>
<td><code>string</code></td>
<td>${params.outdir}/trees</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<h2>Other parameters</h2>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Type</th>
<th>Default</th>
<th>Required</th>
<th>Hidden</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>python_venv_path</code></td>
<td></td>
<td><code>string</code></td>
<td>/home/mdstengl/2023_1_issues/python_venv/</td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>script_dir</code></td>
<td></td>
<td><code>string</code></td>
<td>/home/mdstengl/2023_1_issues/scripts</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>initial_fastqc_dir</code></td>
<td></td>
<td><code>string</code></td>
<td>./results/initial_fastqc/</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>post_trim_fastqc_dir</code></td>
<td></td>
<td><code>string</code></td>
<td>./results/post_trim_fastqc/</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>trimmed_outdir</code></td>
<td></td>
<td><code>string</code></td>
<td>./results/trimmed_fastq</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>blast_outdir</code></td>
<td></td>
<td><code>string</code></td>
<td>./results/blast</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>refseq_dir</code></td>
<td></td>
<td><code>string</code></td>
<td>/home/mdstengl/2023_1_issues/refseq/</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>singularity_pull_docker_container</code></td>
<td></td>
<td><code>boolean</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>metadata</code></td>
<td></td>
<td><code>string</code></td>
<td>None</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>publish_dir_mode</code></td>
<td></td>
<td><code>string</code></td>
<td>link</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>multiqc_title</code></td>
<td></td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>multiqc_config</code></td>
<td></td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>multiqc_logo</code></td>
<td></td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>max_multiqc_email_size</code></td>
<td></td>
<td><code>string</code></td>
<td>25.MB</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>multiqc_methods_description</code></td>
<td></td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>email</code></td>
<td></td>
<td><code>string</code></td>
<td>mark.stenglein@colostate.edu</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>email_on_fail</code></td>
<td></td>
<td><code>string</code></td>
<td>mark.stenglein@colostate.edu</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>plaintext_email</code></td>
<td></td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>monochrome_logs</code></td>
<td></td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>hook_url</code></td>
<td></td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>version</code></td>
<td></td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>validate_params</code></td>
<td></td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>show_hidden_params</code></td>
<td></td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>python_requirements</code></td>
<td></td>
<td><code>string</code></td>
<td>/home/mdstengl/2023_8_dsl2_dev/lib/requirements.txt</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>QC_and_summary_stats</code></td>
<td></td>
<td><code>string</code></td>
<td>results/QC_and_summary_stats</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>cutadapt_trim_reports</code></td>
<td></td>
<td><code>string</code></td>
<td>results/QC_and_summary_stats/cutadapt_trim_reports</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>multiQC_reports</code></td>
<td></td>
<td><code>string</code></td>
<td>results/QC_and_summary_stats/multiqc</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
